packages/boot/src/booters/controller.booter.ts

```diff
@@ -0,0 +1,133 @@
+
+import {CoreBindings, Application, Booter, BootOptions} from '@loopback/core';
+import {inject} from '@loopback/context';
+import {BootBindings} from '../keys';
+import {relative} from 'path';
+import * as util from 'util';
+
+const promisify = util.promisify || require('util.promisify/implementation');
+const glob = promisify(require('glob'));
+
+/**
+ * ControllerBooter is a class that implements the Booter inferface. The purpose
+ * of this booter is to allow for convention based booting of `Controller`
+ * artifacts for LoopBack 4 Applications.
+ *
+ * It supports the following boot phases: config, discover, boot
+ */
+export class ControllerBooter implements Booter {
+  options: ControllerOptions;
+  projectRoot: string;
+  dirs: string[];
+  extensions: string[];
+  discovered: string[];
+
+  /**
+   *
+   * @param app Application instance
+   * @param bootConfig Config options for boot
+   */
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) public app: Application,
+    @inject(BootBindings.BOOT_CONFIG) public bootConfig: BootOptions,
+  ) {
+    if (!bootConfig.controllers) bootConfig.controllers = {};
+    this.options = bootConfig.controllers;
+    this.projectRoot = bootConfig.projectRoot;
+  }
+
+  /**
+   * This phase is responsible for configuring this Booter. It converts
+   * options and assigns default values for missing values so ther phases
+   * don't have to perform checks / conversions.
+   */
+  async configure() {
+    this.options = Object.assign({}, ControllerDefaults, this.options);
+
+    this.dirs = Array.isArray(this.options.dirs)
+      ? this.options.dirs
+      : [this.options.dirs];
+
+    this.extensions = Array.isArray(this.options.extensions)
+      ? this.options.extensions
+      : [this.options.extensions];
+  }
+
+  /**
+   * This phase is responsible for discovering artifact files based on the
+   * given parameters. Sets options.discovered to an array of discovered
+   * artifact files.
+   */
+  async discover() {
+    // glob pattern
+    let pattern = `/@(${this.dirs.join('|')})/`;
+    pattern += this.options.nested ? '**/*' : '*';
+    pattern += `@(${this.extensions.join('|')})`;
+
+    this.discovered = await glob(pattern, {root: this.projectRoot});
+  }
+
+  /**
+   * This phase is responsible for reading the discovered files, checking the
+   * Classes exported and binding them to the Application instance for use.
+   *
+   * It will skip any files it isn't able to load and throw an error containing
+   * a list of skipped files. Other files that were read will still be bound
+   * to the Application instance.
+   */
+  async load() {
+    let errFiles: string[] = [];
+    this.discovered.forEach(file => {
+      try {
+        const ctrl = require(file);
+        const classes: string[] = Object.keys(ctrl);
+        classes.forEach(cls => {
+          this.app.controller(ctrl[cls]);
+        });
+      } catch (err) {
+        errFiles.push(relative(this.projectRoot, file));
```

Comment:
I think we should print the error message to the user. How else are they going to figure out why a file cannot be loaded?

Reply:
Good point. My intention was to continue loading other files if one errored but would it be better to stop and throw the error at that point?

Reply:
I am not sure. Sometimes it is good to see all errors at once, so that one does not have to repeatedly start the app to find all problems. OTOH, if there are dependencies, then the first error can cause a cascade of errors in depending parts of the app, in which case these subsequent errors won't make much sense to the user.

I don't mind either way.

If I had to decide, I would abort on the first error, simply because it's easier to implement.